Source/Core/Core/IOS/Network/Socket.cpp

```diff
@@ -712,24 +717,82 @@ void WiiSockMan::Update()
       socket_iter = WiiSockets.erase(socket_iter);
     }
   }
-  s32 ret = select(nfds, &read_fds, &write_fds, &except_fds, &t);
 
-  if (ret >= 0)
+  // ret is unused as every socket gets updated anyway
+  s32 ret = poll(ufds.data(), nfds, 0);
+  if (ret < 0)
+    ERROR_LOG(IOS_NET, "SO_POLL poll failed (%d)", GetNetErrorCode(ret, "Update", true));
+
+  std::chrono::time_point<std::chrono::high_resolution_clock> now =
+      std::chrono::high_resolution_clock::now();
+  auto elapsed_d = std::chrono::duration_cast<std::chrono::milliseconds>(now - last_time);
+  auto elapsed = elapsed_d.count();
+  last_time = now;
+
+  for (auto& pcmd : pending_polls)
   {
-    for (auto& pair : WiiSockets)
+    if (pcmd.timeout > 0)  // don't touch negative timeouts
     {
-      WiiSocket& sock = pair.second;
-      sock.Update(FD_ISSET(sock.fd, &read_fds) != 0, FD_ISSET(sock.fd, &write_fds) != 0,
-                  FD_ISSET(sock.fd, &except_fds) != 0);
+      if (pcmd.timeout < elapsed)
+        pcmd.timeout = 0;
+      else
+        pcmd.timeout -= elapsed;
     }
   }
-  else
+
+  int n = 0;
+  for (auto& elem : WiiSockets)
   {
-    for (auto& elem : WiiSockets)
+    elem.second.Update(!!(ufds[n].revents & read_event), !!(ufds[n].revents & write_event),
+                       !!(ufds[n].revents & error_event));
+    n++;
+  }
+
+  auto mid = std::partition(pending_polls.begin(), pending_polls.end(), [&](auto& pcmd) {
```

Comment:
I just replaced that with an equivalent [erase-remove idiom](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom).
It still allowed me to remove these two member variables. The only performance benefit to doing it this way instead of iterating is to do only one erase operation, which doesn't really matter in typical cases (IOS doesn't allow polling more than a couple tens of file descriptors). The other benefits in using the stdlib for container manipulations are purely theoretical, but are (maybe) still there. But mostly being less error prone.

Reply:
My only concern about doing it this way, is that it will call `begin()`/`end()`/`remove_if()`/`erase()` every time regardless of the `pending_polls` size. I also think that the whole update poll logic should belong to its own method: "UpdatePollCommands" or something like that.

Reply:
Most (if not all) of these calls are inlined, the only external reference (on g++) is a realloc_insert call, probably for the push_back on ufds, which will probably never get called since its size is preallocated before the loop. I'll move all that logic into a separate function, though, it's true that it's too big to leech off another method.

![Screenshot from 2020-05-19 19-28-38](https://user-images.githubusercontent.com/8816755/82358627-10787600-9a07-11ea-8f9d-0a971eff79f4.png)